user-service/model/user-orm.js

```diff
@@ -1,14 +1,84 @@
-import { createUser } from './repository.js';
+import { createUser, checkUser, deleteUser, updateUser, blackListToken } from './repository.js';
+import jwt from 'jsonwebtoken'
+import bcrypt from 'bcrypt'
+
+const generateJWT = (user) => {
+    return jwt.sign({
+        username: user.username
+    }, process.env.SECRET_TOKEN, { expiresIn: process.env.TOKEN_EXPIRY })
+}
 
 //need to separate orm functions from repository to decouple business logic from persistence
 export async function ormCreateUser(username, password) {
     try {
-        const newUser = await createUser({username, password});
-        newUser.save();
-        return true;
+        const user = await checkUser({username});
+        if (user.length === 0) {
+            const salt = await bcrypt.genSalt()
+            password = await bcrypt.hash(password, salt)
+            const newUser = await createUser({username, password});
+            newUser.save();
+            return newUser;     
+        } else {
+            throw new Error("Username already exist in database");
+        }
     } catch (err) {
         console.log('ERROR: Could not create new user');
         return { err };
     }
 }
 
+export async function ormDeleteUser(token) {
+    try {
+        const { username } = jwt.decode(token)
+        await deleteUser({username})
+        const blacklistToken = await blackListToken({token})
+        blacklistToken.save()
+        return true
+    } catch (err) {
+        console.log("ERROR: Could not delete user")
+        return { err }
+    }
+}
+
+export async function ormUpdateUser(token, password) {
+    try {
+        const { username } = jwt.decode(token)
+        const salt = await bcrypt.genSalt()
+        password = await bcrypt.hash(password, salt)
+        await updateUser({username, password})

```

Comment:
What if the password isn't given? Must we always have a username and password in the request body?

Reply:
Yes, the api will return missing username/password if either one is not given

Reply:
For hashing of password in general, I think it can be done using UserModelSchema.pre("save", () => { "do hashing here"}).

This will ensure that all the password that is saved into the database will be hash. So you don't have to keep manually hash it before saving.

Reply:
ok i will try and update

Reply:
@andrea-twl Do take note.